modules/dynamic/test/src-jvm/smithy4s/dynamic/model/DynamicModelSpec.scala

```diff
@@ -0,0 +1,177 @@
+                    )
+                  )
+                )
+              )
+            )
+          )
+        ),
+        IdRef("foo#Output") -> Shape.StructureCase(
+          StructureShape(
+            members = Some(
+              Map(
+                "greeting" -> MemberShape(
+                  IdRef("smithy.api#String")
+                )
+              )
+            )
+          )
+        )
+      )
+    )
+  }
+
+  test("Decode json representation of models") {
+    IO(
+      SModel
+        .assembler()
+        .addUnparsedModel("foo.smithy", modelString)
+        .assemble()
+        .unwrap()
+    ).map(ModelSerializer.builder().build.serialize(_))
+      .map(NodeToDocument(_))
+      .map(smithy4s.Document.decode[smithy4s.dynamic.model.Model](_))
+      .flatMap(_.liftTo[IO])
+      .map(obtained => expect.same(obtained, expected))
+  }
+
+  object Interpreter {
+    type ToFieldNames[A] = () => List[String]
+
+    object GetFieldNames extends smithy4s.StubSchematic[ToFieldNames] {
+      def default[A]: ToFieldNames[A] = () => Nil
+
+      override def withHints[A](
+          fa: ToFieldNames[A],
+          hints: Hints
+      ): ToFieldNames[A] = fa
+
+      override def genericStruct[S](
+          fields: Vector[Field[ToFieldNames, S, _]]
+      )(const: Vector[Any] => S): ToFieldNames[S] = () =>
+        fields.flatMap { f =>
+          f.label :: f.instance()
+        }.toList
+
+      override def suspend[A](f: => ToFieldNames[A]): ToFieldNames[A] =
+        () => f()
+
+      // these will be needed later but are irrelevant for now
+      // override def union[S](
+      //     first: Alt[ToFieldNames, S, _],
+      //     rest: Vector[Alt[ToFieldNames, S, _]]
+      // )(total: S => Alt.WithValue[ToFieldNames, S, _]): ToFieldNames[S] =
+      //   () =>
+      //     first.label :: first.instance() ::: rest.flatMap { a =>
+      //       a.label :: a.instance()
+      //     }.toList
+
+      // override def list[S](fs: ToFieldNames[S]): ToFieldNames[List[S]] = fs
+      // override def vector[S](fs: ToFieldNames[S]): ToFieldNames[Vector[S]] = fs
+      // override def map[K, V](
+      //     fk: ToFieldNames[K],
+      //     fv: ToFieldNames[V]
+      // ): ToFieldNames[Map[K, V]] = () => fk() ++ fv()
+      // override def bijection[A, B](
+      //     f: ToFieldNames[A],
+      //     to: A => B,
+      //     from: B => A
+      // ): ToFieldNames[B] = f
+
+      // override def set[S](fs: ToFieldNames[S]): ToFieldNames[Set[S]] = fs
+
+    }
+
+    def toFieldNames[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _]](
+        svc: Service[Alg, Op]
+    ): List[String] =
+      svc.endpoints.flatMap { endpoint =>
+        endpoint.input.compile(GetFieldNames)() ++
+          endpoint.output.compile(GetFieldNames)()
+      }
+  }
+
+  pureTest(
+    "Extract field names from all structures in a service's endpoints"
+  ) {
+    val svc = dynamic.Compiler.compile(expected).allServices.head
```

Comment:
The test fails because the smithy prelude is not dumped in the json representation. We would prepopulate the visitor by running it on a handcrafted model containing just the primitives

Reply:
makes sense. I'll let you know if I get to working on it first :)

Reply:
Tried that, turns out the `floatShape` visitor was not doing anything in particular - still the case for other cases. E.g. when I tried to compile PizzaSpec's model, it's doesn't recognize `Food` (a union). I can look at this again after the weekend.

Reply:
> Tried that, turns out the floatShape visitor was not doing anything in particular - still the case for other cases. E.g. when I tried to compile PizzaSpec's model, it's doesn't recognize Food (a union). I can look at this again after the weekend.

Yeah, I had left a fair amount there as TODOs

Reply:
Most of it should be implemented in eb05278